#34, use rufus-scheduler for more flexible scheduling

hummingV opened the pull request:
Use rufus scheduler to support more flexible scheduling options.

It supports four different types of scheduling - cron, every, in and at.
"every" option (e.g. every 5 minutes) offers the same functionality
as existing Stud/interval scheduling. Therefore, 'interval' config
option is deprecated.

1) bumped to 3.1.0

2) Made the following interface changes to the plugin configs:
 + deprecated "interval" option
 + added "schedule" option. e.g schedule => { every => "2m"}

3) If both are specified, raise ConfigurationError.

4) Use existing Stud library if interval option is specified. Use
rufus-scheduler if schedule option is specified.

5) The first invoke of "every" schedule type is configured to run
immediately to match the semantic of interval option.

6) The first invoke of "every" schedule type in rufus-scheduler(v3.0.9)
sometimes does not run immediately, even if :first => :now is specified.
i.e sometimes the first run is invoked immediately after schedule is
started, but sometimes only after the first interval is elasped. The
issue is no longer reproducible in current version (v3.2.2). However,
I can't upgrade to v3.2.2 immediately since other logstash plugins
are using old version. So, I added a workaround that works for both
versions.

karmi commented:
Hi @hummingV, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in yout Git [commit](https://github.com/logstash-plugins/logstash-input-http_poller/pull/58.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

jordansissel commented:
Referencing commit hashes doesn't help for reviewing if we need to review
this later -- can you update this PR with a summary of the change?

On Sunday, July 24, 2016, hummingV <notifications@github.com> wrote:

> See following commit messages for summary of changes
> 992a0f3
> <https://github.com/logstash-plugins/logstash-input-http_poller/commit/992a0f3044321897b5fe00bafcc30f225ee96d97>
> 10550fb
> <https://github.com/logstash-plugins/logstash-input-http_poller/commit/10550fbe3bd817c08978a75a9916d6f9b44a9619>
>
> P.S: kaung and hummingV are both my handles
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/logstash-plugins/logstash-input-http_poller/pull/58
> Commit Summary
>
>    - Use rufus scheduler for more flexible scheduling options
>    - Properly kill the scheduled thread. Improve descriptions.
>    - Avoid unbounded dependency
>    - Avoid breaking change by keeping 'interval' config option
>
> File Changes
>
>    - *M* lib/logstash/inputs/http_poller.rb
>    <https://github.com/logstash-plugins/logstash-input-http_poller/pull/58/files#diff-0>
>    (64)
>    - *M* logstash-input-http_poller.gemspec
>    <https://github.com/logstash-plugins/logstash-input-http_poller/pull/58/files#diff-1>
>    (2)
>    - *M* spec/inputs/http_poller_spec.rb
>    <https://github.com/logstash-plugins/logstash-input-http_poller/pull/58/files#diff-2>
>    (156)
>
> Patch Links:
>
>    -
>    https://github.com/logstash-plugins/logstash-input-http_poller/pull/58.patch
>    -
>    https://github.com/logstash-plugins/logstash-input-http_poller/pull/58.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/logstash-plugins/logstash-input-http_poller/pull/58>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIC6nurSDx45bldBI9vfYxdHnuRwGTvks5qY93NgaJpZM4JTrWK>
> .
>

hummingV commented:
Updated the PR summary.

andrewvc commented on lib/logstash/inputs/http_poller.rb:
```diff
@@ -133,11 +157,35 @@ def validate_request!(url_or_spec, request)
-    Stud.interval(@interval) do
-      run_once(queue)
+    @scheduler = Rufus::Scheduler.new(:max_work_threads => 1)
+    if @interval #interval option is checked first to give it higher precedence
```
A `LogStash::ConfigurationError` should be raised if two schedule options are used by the user at the same time.

This could be DRYed up with code like:

```ruby
raise LogStash::ConfigurationError, "..." if @schedule.keys > 1
schedule_type = @schedule.keys.first
schedule_value = @schedule[schedule_type]
schedule_types = %w(cron every at in)
raise LogStash::ConfigurationError, "..." unless schedule_types.include?(schedule_type)

@scheduler.send(schedule_type, @schedule_value) { run_once(queue) }
```

hummingV commented:
@andrewvc, I've pushed suggested improvements. 

Sorry for the two commits which should have been just one.

andrewvc commented:
Looks good to me! Can you please bump the major version of this plugin, add a changelog entry, and squash the PR down to one commit? 

Once we have that we can merge and release!

Thanks for all the hard work :)

hummingV commented:
@andrewvc, sorry for late reply. Current major version 3.0.3 is not yet released. Do you really want to bump it to 4?

andrewvc commented:
Yes, versions are free! Bumping it to 4 is great :)

hummingV commented:
@andrewvc, it is done. 

I have another request. Our team at Adnovum want to use this feature enhancement on current stable Logstash version (2.3 as of now). Will you be willing to release this feature for current Logstash version as well? Since master branch is on plugin-api-v2, that means you need to release from plugin-api-v1 branch. If fine for you, I will raise another PR on api-v1 branch. --Thanks

suyograo commented:
@hummingV @andrewvc I don't think we should bump version to 4.0. We only bump to major version if there is a backward incompatible change. Looking at the code, there is none for this PR. We could just make it 3.1.0 which means a feature release.

Also @hummingV , yes we can release a version for 2.3.x. Please prep a backport to `plugin-api-v1` branch and we'll merge and push it for you.

hummingV commented:
Ok. I changed the version to 3.1.0 instead. I'd defer versioning decision to you guys. If 3.1.0 is still not ok, please feel free to change the version to whichever that works. 

@suyograo, thanks. I will prepare a backport in a few days.

suyograo commented:
@hummingV this PR also needs a rebase.

hummingV commented:
@suyograo, I don't understand what you mean. I have squashed all changes into one commit. Is there anything more that needs doing?

suyograo commented:
@hummingV there seem to be other changes that got snuck in while this PR is being reviewed. On your local master `git fetch origin` and `git merge origin/master`. You then switch to your `rufus-scheduler` branch and then do `git rebase master` and resolve conflicts..

suyograo commented:
@hummingV You see the green merge button is greyed out...

suyograo commented:
@hummingV tests are failing. Can you look?

hummingV commented:
@suyograo, issue is fixed. 

The first invoke of "every" schedule type turns out to be unpredictable. Sometimes, it runs immediately but sometimes only after the first interval is elapsed. I added a workaround to fix that issue. 

To ensure further that this issue is not immediately exposed to people who were using existing (deprecated with this PR) 'interval' option, I reverted to Stud library for that option.

suyograo commented:
LGTM. Nice work @hummingV ! Thanks a lot for your contribution!
